[Tests not working] Make cli-ui compatible with JRuby.

expeehaa opened the pull request:
As I noted in issue #159, the gem does not work with JRuby currently. In this branch, I tried to change this.
As far as I can tell, this worked, but I couldn’t find a way to replace `fork` in the tests with a working solution. The last of the three commits currently contains my naive idea of making it work, but it doesn’t yet. I’m open for ideas.

The first commit removes the gem `byebug`, because it is not compatible with JRuby. The second commit solves the issue described in #159. This commit alone would make cli-ui work with JRuby, although there are no working tests to confirm this.

lugray commented:
I'm not interested in dropping byebug - we find it useful in our development. Your test changes have, as you've noted, broken testing, so that would need fixing, and finally, commit 2 looks like it would work well for macos and linux, but at a guess it would break down for windows, based on the https://github.com/rdp/os gem you linked in the issue. We'd only be able to bring in jruby support if you were able to resolve these issues.  I'd also want to see jruby add in the `.travis.yml`.

expeehaa commented:
Since creating the PR I haven’t really needed this gem anymore, until maybe now.
I rebased my branch onto the current master. Here is a list of changes compared to my previous branch.
* `byebug` is now only installed when `bundler` recognizes the platform as MRI. It will only be required in tests if the `RUBY_ENGINE` is not JRuby. This is a bit of a conditional mismatch, but I didn’t find a parameter for excluded platforms that could be used in the Gemfile, and `byebug`s README states that it requires MRI anyway.
* OS detection basically uses the same conditions as `os`’s methods `OS.windows?`, `OS.linux?` and `OS.mac?`. If that gem is correct, it should work on Windows now.
* TravisCI runs the tests with `jruby-9.2.14.0` (current version is 9.2.17.0, but `rvm list known` on stable RVM lists 9.2.14.0 as the highest version).
* I tried to make the tests work with jruby, mainly by using `Thread.new` instead of `fork`, but they abort somewhere and I already spent too much time trying to find the cause, so I will not do it in the near future (even if the PR would stay open for another few months). I’ld probably rather rewrite the whole `test/cli/ui/prompt_test.rb` instead of debugging it.
* The tests work with MRI.

lugray commented:
I significantly rewrote the tests in https://github.com/Shopify/cli-ui/compare/expeehaa/add_jruby_compatibility?expand=1 to avoid `fork`.  It still fails on jruby for things involving interrupts, but the rest passes using the same strategy for mri and jruby.  You're welcome to bring in that commit here.  I'll leave understanding the interrupts problem to you, as I'm not familiar with jruby.

expeehaa commented:
Thanks for the test rewriting!

Regarding the interrupt handling I found that the JVM is shut down on SIGINT without ruby rescue code being executed ([see here](https://github.com/jruby/jruby/wiki/Signal-Handling)). I changed the tests to expect the rescue message only if the `RUBY_ENGINE` is not JRuby.
The filename completion test fails because JRuby’s `readline` does not have `Readline::FILENAME_COMPLETION_PROC`, which is used here:
https://github.com/Shopify/cli-ui/blob/4766e789e8170ad1b4f49ac6e8121f43e383a1c4/lib/cli/ui/prompt.rb#L253
I replaced it with a custom proc that should, at least according to my testing, produce the exact same results as that missing constant, but I did not add any additional tests to verify that. I also don’t particularly like the new completion proc code.

The only problems left are spinner tests that seem to be flaky. I’ll try to look into it later this week.

lugray commented:
I reworked some of what you did, and cleaned up our spinner tests. I think this should be good to go now.  Just want to check that you're happy with this, and ask you to sign the [Contributor License Agreement](https://cla.shopify.com/) so that we can include your code in the project.  Thanks for taking an interest and putting in the work to get this going.

As an aside, while you may not have been happy with your filename completion code, it looks good to me.  It gets the job done nicely.

expeehaa commented:
Great, the CLA is signed and I’m happily waiting for the new release now.

lugray commented:
It's been released: https://rubygems.org/gems/cli-ui/versions/1.5.0
